cilium/cmd/fqdn.go

```diff
@@ -43,31 +50,83 @@ var fqdnCleanCacheCmd = &cobra.Command{
 	},
 }
 
-var fqdnCacheClearMatchPattern string
+var fqdnListCacheCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List fqdn cache contents",
+	Run: func(cmd *cobra.Command, args []string) {
+		listFQDNCache()
+	},
+}
+
+var fqdnCacheMatchPattern string
 
 func init() {
+	fqdnCacheCmd.AddCommand(fqdnListCacheCmd)
 	fqdnCacheCmd.AddCommand(fqdnCleanCacheCmd)
 	fqdnCmd.AddCommand(fqdnCacheCmd)
 	rootCmd.AddCommand(fqdnCmd)
 
 	fqdnCleanCacheCmd.Flags().BoolVarP(&force, "force", "f", false, "Skip confirmation")
-	fqdnCleanCacheCmd.Flags().StringVarP(&fqdnCacheClearMatchPattern, "pattern", "p", "", "Cache entries with fqdn matching the pattern will be deleted")
+	fqdnCleanCacheCmd.Flags().StringVarP(&fqdnCacheMatchPattern, "pattern", "p", "", "Cache entries with fqdn matching the pattern will be deleted")
+
+	fqdnListCacheCmd.Flags().StringVarP(&fqdnCacheMatchPattern, "pattern", "p", "", "List cache entries with FQDN matching the pattern")
+	command.AddJSONOutput(fqdnListCacheCmd)
 }
 
 func cleanFQDNCache() {
-	if !force && !confirmCleanup() {
-		return
+	if !force {
+		fmt.Println("Following cache entries are going to be deleted:")
+		listFQDNCache()
+		if !confirmCleanup() {
+			return
+		}
 	}
 
 	params := policy.NewDeleteFqdnCacheParams()
 
-	if fqdnCacheClearMatchPattern != "" {
-		params.SetMatchpattern(&fqdnCacheClearMatchPattern)
+	if fqdnCacheMatchPattern != "" {
+		params.SetMatchpattern(&fqdnCacheMatchPattern)
 	}
 
 	_, err := client.Policy.DeleteFqdnCache(params)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+		return
 	}
 	fmt.Println("FQDN proxy cache cleared")
 }
+
+func listFQDNCache() {
+	params := policy.NewGetFqdnCacheParams()
+
+	if fqdnCacheMatchPattern != "" {
+		params.SetMatchpattern(&fqdnCacheMatchPattern)
+	}
+
+	var lookups []*models.DNSLookup = []*models.DNSLookup{}
+
+	result, err := client.Policy.GetFqdnCache(params)
+	if err != nil {
+		switch err := err.(type) {
+		case *policy.GetFqdnCacheNotFound:
+			// print out empty lookups slice
+		default:
+			fmt.Fprintf(os.Stderr, "Error: %s\n", err)
+			return
+		}
+	} else {
+		lookups = result.Payload
+	}
+
+	if command.OutputJSON() {
+		if err := command.PrintOutput(lookups); err != nil {
+			Fatalf("Unable to provide JSON output: %s", err)
+		}
+	} else {
+		for _, lookup := range lookups {
+			fmt.Printf("%s -> (%s) from %d until %s\n",
+				lookup.Fqdn, strings.Join(lookup.Ips, ", "),
+				lookup.EndpointID, lookup.ExpirationTime.String())
```

Comment:
This was generally unclear... I didn't even realise that the (1234) was an endpoint ID. I'd say we should go with the tabbed columns, something like:
```
Endpoint	FQDN	LookupTime	TTL	ExpirationTime	IPs
1234	cilium.io.	2019-02-05T09:55:37.504Z	60	2019-02-05T09:56:37.504Z	104.198.14.52 2a02:1203:ecb1:b710:b2ea:bcff:fe13:b150
```

I'm pretty sure that will run too wide, though, so we can ditch the LookupTime column. I comma-separated the IPs to make basic CLI parsing a bit simpler, but spaces would also work if that is more readable (or anything but a tab, really).

Reply:
done